python/sglang/srt/models/llama4.py

```diff
@@ -105,20 +105,66 @@ def __init__(
             reduce_results=False,  # We need to do scatter before reduce
         )
 
-    def forward(self, hidden_states):
+    def forward(self, hidden_states, forward_batch: ForwardBatch):
+        shared_out, routed_out = self._forward_core(
+            hidden_states, forward_batch.forward_mode
+        )
+
+        out_aD = routed_out + shared_out
+
+        if self.tp_size > 1:
+            out_aD = tensor_model_parallel_all_reduce(out_aD)
+
+        return out_aD
+
+    def _forward_core(self, hidden_states, forward_mode: ForwardMode):
+        if self._enable_shared_routed_overlap(hidden_states, forward_mode):
+            return self._forward_core_shared_routed_overlap(hidden_states)
+        else:
+            return self._forward_core_normal(hidden_states)
+
+    @staticmethod
+    def _enable_shared_routed_overlap(hidden_states, forward_mode: ForwardMode):
+        batch_size, _ = hidden_states.shape
+        return forward_mode.is_decode() and (batch_size < 4)
```

Comment:
nit: Add a comment explaining why it is enabled only in the `forward_mode.is_decode()` mode and when `batch_size < 4`.